### PR TITLE
JAMES-3925 Rework JMAP quota cleanup

### PR DESCRIPTION
**Pull request by vttranlina**
jira https://issues.apache.org/jira/browse/JAMES-3925

### CODE REVIEW

File: server/container/guice/cassandra/src/main/java/org/apache/james/modules/data/CassandraJmapModule.java [hunk 1]
```diff
@@ -80,7 +79,6 @@ protected void configure() {
         bind(CassandraUploadRepository.class).in(Scopes.SINGLETON);
         bind(UploadDAO.class).in(Scopes.SINGLETON);
         bind(UploadRepository.class).to(CassandraUploadRepository.class);
-        bind(UploadConfiguration.class).toInstance(UploadConfiguration.SINGLETON);
```
Comment (chibenwa) [inline server/container/guice/cassandra/src/main/java/org/apache/james/modules/data/CassandraJmapModule.java:83]: Modify documentation?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/UploadDAO.java [hunk 2]
```diff
@@ -183,13 +172,18 @@ public UploadDAO(CqlSession session, BlobId.Factory blobIdFactory, UploadConfigu
             .whereColumn(USER).isEqualTo(bindMarker(USER))
             .whereColumn(ID).isEqualTo(bindMarker(ID))
             .build());
+
+        this.listByUploadDateLessThan = session.prepare(selectFrom(TABLE_NAME)
+            .all()
+            .allowFiltering()
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/UploadDAO.java:178]: Why?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/UploadDAO.java [hunk 3]
```diff
@@ -183,13 +172,18 @@ public UploadDAO(CqlSession session, BlobId.Factory blobIdFactory, UploadConfigu
...
+            .whereColumn(UPLOAD_DATE).isLessThan(bindMarker(UPLOAD_DATE))
```
Comment (chibenwa) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/UploadDAO.java:179]: Do not push the where clause to the DB and prefer doing app side filtering IMO

**Comment by chibenwa** [comment]
[Test Result](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1701/1/testReport/) (1 failure )

    [org.apache.james.jmap.cassandra.upload.CassandraUploadRepositoryTest.uploadShouldExpire](https://ci-builds.apache.org/job/james/job/ApacheJames/job/PR-1701/1/testReport/junit/org.apache.james.jmap.cassandra.upload/CassandraUploadRepositoryTest/uploadShouldExpire/)

File: server/container/guice/cassandra/src/main/java/org/apache/james/modules/data/CassandraJmapModule.java (hunk 1 above)
Comment (vttranlina) [reply server/container/guice/cassandra/src/main/java/org/apache/james/modules/data/CassandraJmapModule.java:83]: No documentation related to upload configuration (it was not added before)

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java [hunk 4]
```diff
@@ -43,6 +44,9 @@
 import reactor.core.publisher.Mono;
 
 public class CassandraUploadRepository implements UploadRepository {
+
+    public static final BucketName UPLOAD_BUCKET = BucketName.of("uploads_2");
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:48]: Can't find a better name than that for a bucket? `jmap_uploads` for example?

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java [hunk 5]
```diff
@@ -43,56 +44,56 @@
...
     }
 
     public Mono<Void> purge() {
-        return Flux.from(blobStore.listBuckets())
-            .<UploadBucketName>handle((bucketName, sink) -> UploadBucketName.ofBucket(bucketName).ifPresentOrElse(sink::next, sink::complete))
-            .filter(bucketNameGenerator.evictionPredicate())
-            .concatMap(bucket -> blobStore.deleteBucket(bucket.asBucketName()))
+        Instant sevenDaysAgo = clock.instant().minus(EXPIRE_DURATION);
+        return Flux.from(uploadDAO.all())
+            .filter(upload -> upload.getUploadDate().isBefore(sevenDaysAgo))
```
Comment (Arsnael) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: From what I understood when grooming, I thought the purge was supposed to take out maybe 50% of user's JMAP uploads when running it, purging the oldest uploads based on upload_date? (conclusion being, when hitting the quota limit for the user, we cleanup 50% of its oldest objects in the store)

I hardly see how we would limit quota here if we don't purge anything younger than 7 days... People could still abuse the object store within that 7 days period, exploding their quota, correct? I don't understand how is this logic much different than the current one in place

Comment (vttranlina) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: >  I don't understand how is this logic much different than the current one in place

for this task scope, I thought we just changed the mechanism and didn't change the logic.
This method is now invoked by webadmin, I don't know if that is also used when uploading quota
Please correct me

Comment (Arsnael) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: The ticket lacks details on that point but from the overall grooming that's how I understood it

I don't see why we should implementing yet an other cleanup task when uploading and reaching quota limit. or should we have some extra parameters? To delete older than 7 days (that seems very arbitrary as a duration by the way) or 50%? 

Maybe @chibenwa can confirm thoughts here

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: Cleanup is less usefull that is for sure.

But if I upload something then go inactive, do I want the docs to remain stored forever?

IMO there is still a use for that cleanup.

Comment (Arsnael) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: Good use case, agreed :) Should we have an option to choose between that and 50% then? Or that would come with the ticket applying the upload quota limit when doing an upload via jmap then?

Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java:94]: IMO both the cleanup and the quota shall apply.

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java [hunk 6]
```diff
@@ -90,7 +90,8 @@ public Flux<UploadMetaData> listUploads(Username user) {
 
     public Mono<Void> purge() {
         Instant sevenDaysAgo = clock.instant().minus(EXPIRE_DURATION);
-        return Flux.from(uploadDAO.listByUploadDateLessThan(sevenDaysAgo))
+        return Flux.from(uploadDAO.all())
+            .filter(upload -> upload.getUploadDate().isBefore(sevenDaysAgo))
             .concatMap(upload -> Mono.from(blobStore.delete(UPLOAD_BUCKET, upload.getBlobId()))
                 .then(uploadDAO.delete(upload.getUser(), upload.getId())))
```
Comment (quantranhong1999) [inline server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java]: Q: Should we use flatMap to fasten the deletion?

**Review by quantranhong1999 — APPROVED** [review]

File: server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java (hunk 6 above)
Comment (chibenwa) [reply server/data/data-jmap-cassandra/src/main/java/org/apache/james/jmap/cassandra/upload/CassandraUploadRepository.java]: A: likely

**Review by hungphan227 — APPROVED** [review]

**Review by chibenwa — APPROVED** [review]

**Review by quantranhong1999 — APPROVED** [review]

**Review by Arsnael — APPROVED** [review]

**Pull request merged**